### PR TITLE
Gb/slurm refactor

### PR DESCRIPTION
**Pull request by grantbuster**
Starting PR process but should do a full test on eagle before we merge.

### CODE REVIEW

File: rex/utilities/hpc.py [hunk 1]
```diff
@@ -0,0 +1,738 @@
...
+                status = self.queue[int(job_id)][self.QCOL_STATUS]
+
+        elif job_name is not None:
+            if job_name in self.queue_job_names:
+                for attrs in self.queue.values():
+                    if attrs[self.QCOL_NAME] == job_name:
+                        status = attrs[self.QCOL_STATUS]
+                        break
+
+        else:
```
Comment (MRossol) [inline rex/utilities/hpc.py:208]: what if job_id and job_name are given?

File: rex/utilities/hpc.py [hunk 2]
```diff
@@ -0,0 +1,738 @@
...
+class PBS(HpcJobManager):
+    """Subclass for PBS subprocess jobs."""
+
+    # PBS qstat column headers
+    QCOL_NAME = 'Name'  # Job name column
+    QCOL_ID = 'Job id'  # Job integer ID column
+    QCOL_STATUS = 'S'  # Job status column
+
+    # Frozen PBS qstat column headers cached b/c its not space delimited
+    QSTAT_KEYS = ('Job id', 'Name', 'User', 'Time Use', 'S', 'Queue')
```
Comment (MRossol) [inline rex/utilities/hpc.py:225]: Since we are defining these, and thoughts about dropping spaces?

**Review by MRossol — CHANGES_REQUESTED** [review]
Instead of defining:
```
    QCOL_NAME = 'Name'  # Job name column
    QCOL_ID = 'Job id'  # Job integer ID column
    QCOL_STATUS = 'S'  # Job status column
```

What do you think about instead overwriting the header by defining `QSKIP` and `QSTAT_KEYS` such that the `name`, `id`, and `status` columns are always the same?

I can see pros and cons to both approaches, so it's just a thought...

File: rex/utilities/hpc.py (hunk 1 above)
Comment (grantbuster) [reply rex/utilities/hpc.py:208]: job id is first in the if statement so it gets priority

File: rex/utilities/hpc.py (hunk 2 above)
Comment (grantbuster) [reply rex/utilities/hpc.py:225]: I think we should leave it how we expect PBS to print it. If the user is interacting with a dictionary of the qstat they would expect the qstat column headers, not some arbitrary ones that we pick

Comment (grantbuster) [reply rex/utilities/hpc.py:225]: Also fyi this is just from some example online. I can't remember what PBS actually prints out and i expect its different from PBS config to config.

Comment (MRossol) [reply rex/utilities/hpc.py:225]: As we said 90% there!

File: rex/utilities/hpc.py (hunk 1 above)
Comment (MRossol) [reply rex/utilities/hpc.py:208]: Sounds good

**Comment by grantbuster** [comment]
@MRossol i think we should leave the queue column headers as we expect PBS / SLURM to return them. I guess it doesnt really matter too much if we enforce our own keys but I think random users might expect to see what they see in PBS / SLURM. It's also nice because if those keys change or there's a new HPC class we make you can just subclass and change the class attrs.

**Review by MRossol — APPROVED** [review]
Assuming tests pass ;-)

**Comment by MRossol** [comment]
Caught a bug on Eagle:
```╰─ reV -c config_batch.json batch
INFO - 2020-10-09 12:24:50,331 [pipeline.py:132] : reV pipeline submitting: "generation" for job "set1_wthh80_wtpp0"
ERROR - 2020-10-09 12:24:53,170 [pipeline.py:140] : Pipeline subprocess submission returned an error:
Subprocess submission failed with return code 1 and stderr:
Error running reV CLI
Traceback (most recent call last):
  File "/home/mrossol/reV/reV/cli.py", line 325, in <module>
    main(obj={})
  File "/home/mrossol/miniconda3/envs/rev/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/mrossol/miniconda3/envs/rev/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/mrossol/miniconda3/envs/rev/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/mrossol/miniconda3/envs/rev/lib/python3.7/site-packages/click/core.py", line 1236, in invoke
    return Command.invoke(self, ctx)
  File "/home/mrossol/miniconda3/envs/rev/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/mrossol/miniconda3/envs/rev/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/mrossol/miniconda3/envs/rev/lib/python3.7/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/mrossol/reV/reV/cli.py", line 66, in generation
    verbose=verbose)
  File "/home/mrossol/miniconda3/envs/rev/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/mrossol/miniconda3/envs/rev/lib/python3.7/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/mrossol/reV/reV/generation/cli_gen.py", line 125, in from_config
    submit_from_config(ctx, name, year, config, i, verbose=verbose)
  File "/home/mrossol/reV/reV/generation/cli_gen.py", line 191, in submit_from_config
    verbose=verbose)
  File "/home/mrossol/miniconda3/envs/rev/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/mrossol/miniconda3/envs/rev/lib/python3.7/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/mrossol/reV/reV/generation/cli_gen.py", line 725, in slurm
    module=module)[0]
  File "/home/mrossol/rex/rex/utilities/hpc.py", line 734, in sbatch
    self._queue[int(out)] = {self.QCOL_ID: int(out),
ValueError: invalid literal for int() with base 10: 'Submitted batch job 4747868'
```

**Comment by grantbuster** [comment]
@MRossol fyi i finished testing on Eagle and everything is running smoothly. It's definitely more performant than rex/master but not by a mile. I think it will be a lot better when the SLURM scheduler is overloaded though.

**Pull request merged**